Configurations/15-android.conf

```diff
@@ -26,7 +26,7 @@
                 my $api = "*";
 
                 # see if user passed -D__ANDROID_API__=N
-                foreach (@{$useradd{CPPDEFINES}}) {
+                foreach (@{$useradd{CPPDEFINES}}, @{$user{CPPFLAGS}}) {
```

Comment:
If you want to cover everything, you should check `@{$user{CPPDEFINES}}` as well...

Reply:
Hmmm, I would actually rather challenge existence of `@{$user{CPPDEFINES}}`. I mean what is `$user{CPPDEFINES}`? It appears like it's a make/environment variable. But who would use it? (And how many even know how?) It doesn't look like something standard. And it's even resolved prior it goes to makefile. I mean it looks like make/env variable, but it doesn't appear in resulting makefile... This is kind of counter-intuitive and sends all wrong signals. Because idea behind make/environment variables is to meet other users' expectations and facilitate integration with other projects. And their expectations are CPPFLAGS and that's it. Well, I suppose you can come up with some useful corner cases, but I don't  see Android users among them...

In the context you can wonder what do I think of `@{$useradd{CPPDEFINES}}`. Well, approximately same, i.e. I think it's excessive, CPPFLAGS should have sufficed. However! It's all internal thing, in sense that it's not something user can assign *directly*. It's assigned indirectly as result of classification of pre-processor-looking command line options. In other words it's an internal implementation detail and not part of user "interface" such as make/env variable. (Though one can still wonder if it's actually appropriate to classify them this way, because when they are reconstructed in CPPFLAGS in makefile they can come out in different order, and it's not actually given that it's universally appropriate, to change the order that is.)

Reply:
`CPPDEFINES` is documented in `INSTALL`.  Granted, its existence is due to how the VMS C compiler flags work, but I made it available for everyone.  So if someone reads the `INSTALL` and thinks `CPPDEFINES` is a good idea, they might use it.

Reply:
> how the VMS C compiler flags work

Or rather how passing arguments in DCL works, but never mind. I'm not really in position to speak for VMS users, but I'd argue that even them would be better off specifying CPPFLAGS and forming additional defines to DCL liking. Well, it might happen that it would be impossible to do so at DCL prompt. I mean it might be impossible to CPPFLAGS=/DEFINE=(A,B,C), because it might confuse DCL into believing that it's two arguments, not one. Is this it?

Reply:
No.  The thing is that the VMS C compiler will only care about the last `/DEFINE` (*apparently, it accepts the `/DEFINE` is given more than once, but will still ignore all but the last...  that came as a surprise when I discovered it*).  That means that between all the macros the user might want to add, together with all the macros *we* generate, they need to be collected together and placed into one `/DEFINE`.  The same goes for `/INCLUDE`.  This is the reason we also have the `defines` and `includes` in the config targets.
I didn't want to have to parse out `/DEFINE` and `/INCLUDE` and all that crap, so my solution was to have separate variables for their content.  This also provisions for compilers on other non-Unix platforms that might come with similar limitations, should we ever encounter another...

Reply:
> I didn't want to have to parse out /DEFINE and /INCLUDE and all that crap,

Then why not **demand** that VMS users pass them with -D and -I make the whole thing an "internal implementation" detail. I'd say that above is rather weak argument. You either make it intuitive to user (by parsing whatever user expects natural), or you impose something on them. And in last case it's more appropriate to impose something that they can relate to based on experience with another system than to invent some interface thing from scratch and impose it on everybody.

Either way, I still stand by "I can't see that Android users would use it."

Reply:
I'm not *imposing* anything on the user here, I'm just saying that there's a choice. Now, if you don't see them using it then fine.

(speaking of imposing, your argument that VMS users should be required to use -D and -I is exactly that)

Reply:
I think that having more than one way to specify things here is not helpful, across all systems.

Reply:
> (speaking of imposing, your argument that VMS users should be required to use -D and -I is exactly that)

But so are you :-) What I'm saying is that as result of refusal to parse out /DEFINE and /INCLUDE (as part of CPPFLAGS) *something* has to be imposed on VMS users. And what I'm saying is that everybody would be better off if we imposed something based on experience with *other* systems they can relate to.

Reply:
Seriously, this discussion is getting off on a tangent.  I'm dropping this request.